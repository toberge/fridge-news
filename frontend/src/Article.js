// @flow

export class ArticleBase {
  id: number;
  title: string;
  picturePath: string;
  pictureAlt: string;
  category: string;
  date: Date;
  rating: number;

  constructor(id: number, title: string, picturePath: string, pictureAlt: string, category: string, date: Date, rating: number) {
    this.id = id;
    this.title = title;
    this.picturePath = picturePath;
    this.pictureAlt = pictureAlt;
    this.category = category;
    this.date = date;
    this.rating = rating;
  }
}

export class Article {
  id: number;
  title: string;
  ingress: string;
  picturePath: string;
  pictureAlt: string;
  pictureCapt: string;
  text: string;
  category: string;
  date: Date;
  author: string; // TODO plz be author obj or somthing
  rating: number;

  constructor(id: number, title: string, ingress: string, picturePath: string, pictureAlt: string, pictureCapt: string, text: string, category: string, date: Date, author: string, rating: number) {
    this.id = id;
    this.title = title;
    this.ingress = ingress;
    this.picturePath = picturePath;
    this.pictureAlt = pictureAlt;
    this.pictureCapt = pictureCapt;
    this.text = text;
    this.category = category;
    this.date = date;
    this.author = author;
    this.rating = rating;
  }
}
