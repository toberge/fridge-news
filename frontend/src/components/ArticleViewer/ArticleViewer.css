article.hidden {
  opacity: 0;
  pointer-events: none;
}

article {
  -webkit-transition: opacity 0.5s;
  -moz-transition: opacity 0.5s;
  -ms-transition: opacity 0.5s;
  -o-transition: opacity 0.5s;
  transition: opacity 0.5s;
  opacity: 1;
}

figure {
  margin: 0;
  text-align: center;
}

figure img {
  display: inline-block;
  text-align: center;
  width: 100%;
}

.dateline {
  background-color: var(--section-background);
  padding: 1rem;
  -webkit-border-radius: 0.3rem;
  -moz-border-radius: 0.3rem;
  border-radius: 0.3rem;
}

@media screen and (min-width: 961px) {
  .dateline {
    text-align: center;
  }

  .dateline dt, .dateline dd {
    display: inline-block;
  }

  .dateline dt {
    margin-right: 0.5rem;
  }

  .dateline dd {
    margin-right: 1rem;
  }
}

.article-buttons {
  text-align: center;
}
