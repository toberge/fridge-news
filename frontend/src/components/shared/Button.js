// @flow

import * as React from 'react';
import { Component } from 'react-simplified/lib/index';

class ButtonPrimary extends Component<{
  onClick: () => mixed,
  children?: React.Node,
  disabled?: boolean
}> {
  render() {
    return (
      <button type="button" className="btn btn-primary" onClick={this.props.onClick} disabled={this.props.disabled}>
        {this.props.children}
      </button>
    );
  }
}

class ButtonSuccess extends Component<{
  onClick: () => mixed,
  children?: React.Node
}> {
  render() {
    return (
      <button type="button" className="btn btn-success" onClick={this.props.onClick}>
        {this.props.children}
      </button>
    );
  }
}

class ButtonLight extends Component<{
  onClick: () => mixed,
  children?: React.Node
}> {
  render() {
    return (
      <button type="button" className="btn btn-light" onClick={this.props.onClick}>
        {this.props.children}
      </button>
    );
  }
}

class ButtonDanger extends Component<{
  onClick: () => mixed,
  children?: React.Node
}> {
  render() {
    return (
      <button type="button" className="btn btn-danger" onClick={this.props.onClick}>
        {this.props.children}
      </button>
    );
  }
}

/**
 * Button widgets
 */
export default class Button {
  static Success = ButtonSuccess;
  static Secondary = ButtonLight;
  static Danger = ButtonDanger;
  static Primary = ButtonPrimary;
}
