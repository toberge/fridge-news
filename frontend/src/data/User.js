// @flow

export default class User {
  id: number;
  name: string;
  admin: boolean;

  constructor(id: number, name: string, admin: boolean) {
    this.id = id;
    this.name = name;
    this.admin = admin;
  }
}
