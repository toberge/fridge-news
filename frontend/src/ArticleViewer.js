// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import MarkdownRenderer from 'react-markdown-renderer';
import { Article } from './Article';
import { articleService } from './Services';

export class ArticleViewer extends Component<{ match: { params: { id: number } } }> {
  render() {
    const {
      title,
      ingress,
      picturePath,
      pictureAlt,
      pictureCapt,
      text,
      author,
      category,
      date,
      rating
    } = articleService.currentArticle;
    return (
      <article>
        <header>
          <h2>{title}</h2>
          <p className="ingress">
            <em>{ingress}</em>
          </p>
        </header>
        <figure>
          <img src={picturePath} alt={pictureAlt} />
          <figcaption>
            <p>
              <em>Image:</em>
              {' ' + pictureCapt}
            </p>
          </figcaption>
        </figure>
        <MarkdownRenderer markdown={text} />
        {/*comments TODO*/}
        <section className="comments">
          <h2>Comments</h2>
          <div className="comment">
            <p>
              <em>
                <span className="author">Grumbling Grevory</span> said:
              </em>
            </p>
            <p>I really hate this article...</p>
          </div>
          <div className="comment">
            <p>
              <em>
                <span className="author">Somebody</span> said:
              </em>
            </p>
            <p>
              Everything is fine.
              <br />
              It really is.
            </p>
          </div>
          <h2>Leave your own comment</h2>
          <p>
            <label htmlFor="comment-text">Comment:</label>
            <br />
            <textarea
              id="comment-text"
              cols="80"
              rows="10"
              aria-label="Comment:"
            ></textarea>
            <br />
            <button id="comment-submit">Submit</button>
          </p>
        </section>
      </article>
    );
  }

  mounted(): void {
    articleService.getArticle();
  }
}
