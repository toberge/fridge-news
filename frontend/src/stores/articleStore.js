// @flow

import axios from 'axios';
import { Article, ArticleBase } from '../data/Article';
import { sharedComponentData } from 'react-simplified';

const placeholder = new Article(
    1,
    1,
    'Fridge Found Floating in Space',
    'https://i.imgur.com/puQs66y.png',
    'Fridge floating in space',
    'The infamous space-floating fridge',
    `*After its long flight through the rings of Saturn, a fridge of mysterious
origin was caught by a spacewalking astronaut on the ISS and brought to
Earth for inspection. Inside were the remains of an unknown biological life
form - and a pile of ancient floppy disks. The floppies contained impressive
amounts of text - newspaper articles, essays, even a few poems. Here are a
few snippets from the fridge's collection.*
### Space Cows Demand We Go Vegan  
Yesterday, a fleet of extrasolar ships approached Earth, bearing an ultimatum.
If we do not stop eating meat and return all our livestock to nature, our entire civilization will be eradicated.
The species behind this harsh request look almost exactly like the cows of humanity's home world, but their intelligence is on a whole other level, perhaps exceeding ours.
Most governments in the solar system have already stated that they perceive this threat as very real, some adding that they will take immediate action to enforce a meat ban.`,
    'news',
    new Date(),
    null,
    1,
    2.3
  );

const empty = new Article(
    1,
    1,
    '',
    '',
    '',
    '',
    `*Begin your article with an ingress*
    
## Use headers of level 2 and below

write some **good** text`,
    '',
    new Date(),
    null,
    2,
    2.3
  );

class ArticleStore {
  // won't make loading multiple articles necessitate showing a loading page,
  // it's better to show a not-yet-updated list than to show *the wrong article*
  loadingArticle: boolean = false;
  currentArticle: Article = placeholder;
  articles: ArticleBase[] = [];
  categoryMap: Map<string, ArticleBase[]> = new Map<string, ArticleBase[]>([
    ['news', []],
    ['culture', []],
    ['science', []],
    ['politics', []]
  ]);

  categories: string[] = ['news', 'culture', 'science', 'politics', 'blablabla'];

  // wow shit why must I be forced to do this...
  getCategories() {
    return axios.get('/articles/categories/')
      .then(response => response.data)
      .then(strings => {
        // replace old array (if anything lives there)
        const countBefore = this.categories.length;
        this.categories.push(strings);
        this.categories.splice(0, countBefore);
        // add to hashmap of cached categories
        strings.forEach(s => {
          if (!this.categoryMap.get(s)) {
            this.categoryMap.set(s, []);
          }
        })
      })
  }

  clearArticle() {
    Object.assign(this.currentArticle, { ...empty });
  }

  getArticle(id: number) {
    this.loadingArticle = true;
    return axios
      .get<Article>('/articles/' + id)
      .then(response => response.data)
      .then(result => {
        const {
          article_id,
          user_id,
          title,
          picture_path,
          picture_alt,
          picture_caption,
          content,
          upload_time,
          update_time,
          importance,
          category,
          rating
        } = result;
        this.currentArticle = new Article(
          article_id,
          user_id,
          title,
          picture_path,
          picture_alt,
          picture_caption,
          content,
          category,
          new Date(Date.parse(upload_time)),
          update_time ? new Date(Date.parse(update_time)) : null,
          importance,
          rating
        );
        this.loadingArticle = false;
        return result;
      });
  }

  addArticle(): Promise<number | void> {
    const article = this.currentArticle;
    return axios.post('/articles/', {
      user_id: article.authorID,
      title: article.title,
      picture_path: article.picturePath,
      picture_alt: article.pictureAlt,
      picture_caption: article.pictureCapt,
      content: article.text,
      importance: article.importance,
      category: article.category
    }).then(response => response.data.id);
  }

  updateArticle(): Promise<boolean | void> {
    const article = this.currentArticle;
    return axios.put('/articles/' + article.id, {
      title: article.title,
      picture_path: article.picturePath,
      picture_alt: article.pictureAlt,
      picture_caption: article.pictureCapt,
      content: article.text,
      importance: article.importance,
      category: article.category
    }).then(response => response.status === 200);
  }

  deleteArticle(): Promise<boolean> {
    return axios.delete(`/articles/${this.currentArticle.id}`)
      .then(response => response.status === 200);
  }

  getFrontPage(): Promise<ArticleBase[]> {
    return axios
      .get<ArticleBase[]>('/articles/front_page')
      .then(response => {
        this.articles.splice(0, this.articles.length);
        this.articles.push(...this.toArticleArray(response.data));
      });
  }

  getCategory(category: string): Promise<ArticleBase[]> {
    return axios
      .get<ArticleBase[]>('/articles/categories/' + category)
      .then(response => {
        const array = this.categoryMap.get(category);
        if (array) {
          array.splice(0, array.length);
          array.push(...this.toArticleArray(response.data));
        }
      })
      /*.catch(e => {
        // this is temporary crap
        this.articles.splice(0, this.articles.length);
        throw e;
      })*/
  }

  toArticleArray(result: []): ArticleBase[] {
    return result.map(e => {
      const {article_id, title, picture_path, picture_alt, category} = e;
      return new ArticleBase(article_id, title, picture_path, picture_alt, category);
    });
  }
}

export const articleStore = sharedComponentData(new ArticleStore());
