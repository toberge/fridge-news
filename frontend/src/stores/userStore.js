// @flow

import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';
import { sharedComponentData } from 'react-simplified';
import User from '../data/User';
import Notifier from '../components/shared/Notifier';

class UserStore {
  // login state
  currentUser: ?User;
  token: ?string;
  tokenInterval: ?IntervalID = null;
  // other users
  currentAuthor: User = new User(1, 'The Fridge', true);
  cachedUsers: Map<number, User> = new Map<number, User>();

  loadUser() {
    const username = sessionStorage.getItem('username');
    const id = sessionStorage.getItem('userID');
    const token = sessionStorage.getItem('token');
    if (username && id && token) {
      // set temp user
      this.currentUser = new User(parseInt(id), username, false);
      this.token = token;
      // load user from DB
      this.getUser(parseInt(id))
        .then(user => this.currentUser = user)
        .catch(error => {
          console.log(error);
          this.currentUser = null;
        });
    }
  }

  saveUser() {
    if (this.currentUser && this.token) {
      sessionStorage.setItem('username', this.currentUser.name);
      // excessive checks because Flow is stupid sometimes
      if (this.currentUser) sessionStorage.setItem('userID', `${this.currentUser.id}`);
      if (this.token) sessionStorage.setItem('token', this.token);
    }
  }

  logIn(name: string, password: string) {
    return axios.post('/login', { name, password }).then(async (response: AxiosResponse) => {
      // already okay
      console.log(response.data);
      this.currentUser = await this.getUser(response.data.user_id);
      this.token = response.data.jwt;
      this.startTokenInterval();
      this.saveUser();
      return true;
    });
  }

  /**
   * Just logs off our dear user
   */
  logOut() {
    this.currentUser = null;
    this.token = null;
    sessionStorage.clear();
    clearInterval(this.tokenInterval);
  }

  register(username: string, password: string) {
    return axios.post('/users/', { name: username, password }).then(async response => {
      this.currentUser = await this.getUser(response.data.insertId);
      this.token = response.data.jwt;
      this.startTokenInterval();
      this.saveUser();
      return this.currentUser ? this.currentUser.id : -1; // TODO temp hack...
    });
  }

  getTokenHeader(): AxiosRequestConfig {
    return {
      headers: {
        'x-access-token': userStore.token
      }
    };
  }

  startTokenInterval() {
    this.tokenInterval = setInterval(this.refreshToken, 1000 * 60 * 4);
  }

  refreshToken() {
    axios
      .get('/token/', this.getTokenHeader())
      .then((response: AxiosResponse) => {
        if (response.data.jwt) this.token = response.data.jwt;
        console.log('fetched token...');
      })
      .catch((e: Error) => {
        Notifier.error(`Refreshing token failed, logging out... Error: ${e.message}`);
        this.logOut();
      });
  }

  getUser(id: number): Promise<User> {
    let user = this.cachedUsers.get(id);
    if (user) {
      return new Promise<User>(resolve => resolve(user));
    } else {
      return axios.get('/users/' + id).then(response => {
        const { user_id, name, admin } = response.data;
        const newUser = new User(user_id, name, admin);
        this.cachedUsers.set(id, newUser);
        return newUser;
      });
    }
  }

  getAuthor(id: number) {
    return this.getUser(id).then(user => (this.currentAuthor = user));
  }
}

export const userStore = sharedComponentData(new UserStore());
