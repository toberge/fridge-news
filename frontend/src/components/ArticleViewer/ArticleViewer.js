// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import MarkdownRenderer from 'react-markdown-renderer';
import 'easymde/dist/easymde.min.css';
import { articleStore } from '../../stores/articleStore';
import './ArticleViewer.css';
import { Button } from '../widgets';
import { createHashHistory } from 'history';
import CommentSection from './CommentSection';
import Notifier from '../shared/Notifier';
import { userStore } from '../../stores/userStore';
import Icon from '../shared/Icon';
import {capitalizeFirstLetter} from "../../data/Article";

const history = createHashHistory();

export default class ArticleViewer extends Component<{ match: { params: { id: number } } }> {
  hidden: boolean = true;

  render() {
    const {
      title,
      picturePath,
      pictureAlt,
      pictureCapt,
      text,
      uploadTime,
      rating
    } = articleStore.currentArticle;
    document.title = `Fridge News | ${title}`;
    return (
      <main>
        {/* TODO split into ArticleRenderer to be used for preview in editor */}
        <article className={this.hidden || articleStore.loadingArticle ? 'hidden' : ''}>
          <header>
            <h1>{title}</h1>
          </header>
          {picturePath && pictureAlt && pictureCapt ? (
            <figure>
              <img src={picturePath} alt={pictureAlt} />
              <figcaption>
                <p>
                  <em>Image:</em>
                  {' ' + pictureCapt}
                </p>
              </figcaption>
            </figure>
          ) : null}
          <MarkdownRenderer markdown={text} />
          <section className="details">
            <dl className="dateline">
              <dt>Author:</dt>
              <dd>{userStore.currentAuthor.name}</dd>
              <dt>Published:</dt>
              <dd>{uploadTime.toLocaleString()}</dd>
              <dt>Rated:</dt>
              <dd>
                <meter value={rating} max={5} min={1} />
                {` (${rating} out of 5)`}
              </dd>
            </dl>
          </section>
          <section className="article-buttons">
            <strong>You may wish to </strong>
            <Button.Secondary onClick={this.handleEdit}>
              <Icon.Write /> Edit
            </Button.Secondary>
            <strong> or </strong>
            <Button.Danger onClick={this.handleDelete}>
              <Icon.Delete /> Delete
            </Button.Danger>
            <strong> this article.</strong>
          </section>
          <CommentSection articleID={this.props.match.params.id} />
        </article>
      </main>
    );
  }

  handleEdit() {
    // TODO prevent this from being necessary?
    if (articleStore.currentArticle.picturePath === null) {
      articleStore.currentArticle.picturePath = '';
      articleStore.currentArticle.pictureAlt = '';
      articleStore.currentArticle.pictureCapt = '';
    }
    history.push(`/articles/${this.props.match.params.id}/edit`);
  }

  handleDelete() {
    articleStore.deleteArticle()
      .then(() => {
        Notifier.success('Successfully deleted article');
        history.goBack(); // TODO handle other page I guess
      })
      .catch((e: Error) => Notifier.error(`Could not delete article\n${e.message}`));
    // TODO maybe don't push immediately
  }

  async mounted() {
    this.hidden = true;
    document.title = 'Loading... - Fridge News';
    try {
      await articleStore.getArticle(this.props.match.params.id);
      await userStore.getAuthor(articleStore.currentArticle.authorID);
      this.hidden = false;
      document.title = `${articleStore.currentArticle.title} - ${capitalizeFirstLetter(articleStore.currentArticle.category)} - Fridge News`
    } catch (e) {
      Notifier.error(`Could not fetch article, reason:\n${e.message}`);
      history.push('/404');
    }
  }
}
