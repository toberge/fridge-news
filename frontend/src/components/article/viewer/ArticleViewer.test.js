// @flow

import * as React from 'react';
import 'core-js/stable';
import 'regenerator-runtime/runtime';
import { shallow, ShallowWrapper } from 'enzyme/build';
import ArticleViewer from './ArticleViewer';
import { articleStore } from '../../../stores/articleStore';
import { userStore } from '../../../stores/userStore';
import User from '../../../data/User';
import { Article } from '../../../data/Article';
import Button from '../../shared/Button';

jest.mock('../../../stores/articleStore');
jest.mock('history');

describe('ArticleViewer', () => {
  // set up data
  articleStore.currentArticle = new Article(
    1,
    2,
    'whatever',
    null,
    null,
    null,
    'blablabla',
    'science',
    new Date('1997-02-23T04:04:04'), // specific dat for snapshot integrity
    null,
    1,
    2.4
  );
  userStore.currentAuthor = new User(2, 'Author of Articles', false);
  userStore.cachedUsers.set(2, userStore.currentAuthor);

  const articleViewer: ShallowWrapper = shallow(<ArticleViewer match={{ params: { id: 1 } }} />);

  const update = () => {
    articleViewer.update();
    articleViewer.instance().forceUpdate();
  };

  it('should fetch article when mounting', () => {
    expect(articleStore.getArticle).toHaveBeenCalled();
  });

  it('should hide controls when user is NOT author of article', () => {
    userStore.currentUser = new User(42, 'white fish that does not write', false);
    update();
    expect(articleViewer.find('.article-buttons')).toHaveLength(0);
    expect(articleViewer.debug()).toMatchSnapshot();
  });

  it('should show controls when user is author of article', () => {
    userStore.currentUser = userStore.currentAuthor;
    update();
    expect(articleViewer.find('.article-buttons')).toHaveLength(1);
    expect(articleViewer.debug()).toMatchSnapshot();
  });

  it('should delete article when corresponding button is pressed', () => {
    userStore.currentUser = userStore.currentAuthor;
    const spy = jest.spyOn(articleStore, 'deleteArticle').mockResolvedValue();
    expect(spy).not.toHaveBeenCalled();
    // diving down to shallowly render button and click it
    articleViewer
      .find(Button.Danger)
      .dive()
      .simulate('click');
    expect(spy).toHaveBeenCalled();
  });

  it('should show picture when the article has one', () => {
    articleStore.currentArticle.picturePath = 'https://bogus.com/image.jÃ¥ddpegg';
    articleStore.currentArticle.pictureAlt = 'intense bogus';
    articleStore.currentArticle.pictureCapt = 'Look at this INTENSITY';
    update();
    expect(articleViewer.find('figure')).toHaveLength(1);
    expect(articleViewer.debug()).toMatchSnapshot();
  });

  it('should not show picture when the article does not have one', () => {
    articleStore.currentArticle.picturePath = null;
    update();
    expect(articleViewer.find('figure')).toHaveLength(0);
    expect(articleViewer.debug()).toMatchSnapshot();
  });
});
