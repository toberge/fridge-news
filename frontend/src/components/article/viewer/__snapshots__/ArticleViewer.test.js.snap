// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleViewer should hide controls when user is NOT author of article 1`] = `
"<main>
  <div className=\\"\\">
    <article>
      <header>
        <h1>
          whatever
        </h1>
      </header>
      <MarkdownRenderer markdown=\\"blablabla\\" />
    </article>
    <aside className=\\"details\\">
      <dl className=\\"dateline\\">
        <dt>
          Author:
        </dt>
        <dd>
          Author of Articles
        </dd>
        <dt>
          Published:
        </dt>
        <dd>
          2/23/1997, 4:04:04 AM
        </dd>
      </dl>
    </aside>
    <CommentSection articleID={1} />
  </div>
</main>"
`;

exports[`ArticleViewer should show controls when user is author of article 1`] = `
"<main>
  <div className=\\"\\">
    <article>
      <header>
        <h1>
          whatever
        </h1>
      </header>
      <MarkdownRenderer markdown=\\"blablabla\\" />
    </article>
    <aside className=\\"details\\">
      <dl className=\\"dateline\\">
        <dt>
          Author:
        </dt>
        <dd>
          Author of Articles
        </dd>
        <dt>
          Published:
        </dt>
        <dd>
          2/23/1997, 4:04:04 AM
        </dd>
      </dl>
    </aside>
    <aside className=\\"article-buttons\\">
      <strong>
        You may wish to 
      </strong>
      <ButtonLight onClick={[Function: bound handleEdit]}>
        <WriteIcon />
         Edit
      </ButtonLight>
      <strong>
         or 
      </strong>
      <ButtonDanger onClick={[Function: bound handleDelete]}>
        <DeleteIcon />
         Delete
      </ButtonDanger>
      <strong>
         this article.
      </strong>
    </aside>
    <CommentSection articleID={1} />
  </div>
</main>"
`;
